### PR TITLE
Addition of KL-Divergences for torch.distributions

### PR DESCRIPTION
**Pull request by vishwakftw**
Math checked and approved by @fritzo .

cc: @apaszke

### CODE REVIEW

File: test/test_distributions.py [hunk 1]
```diff
@@ -1209,23 +1249,99 @@ def test_laplace_shape_tensor_params(self):
...
+            (Gamma(1, 2), Chi2(3)),
+            (Gamma(1, 2), Exponential(3)),
+            (Gamma(1, 2), Gamma(3, 4)),
+            (Gamma(1, 2), Gumbel(-3, 4)),
+            (Gumbel(-1, 2), Gumbel(-3, 4)),
+            (Gumbel(-1, 2), Normal(-3, 4)),  # This case fails for n = 22000
+            (Laplace(1, 2), Laplace(-3, 4)),
+            (Laplace(-1, 2), Normal(-3, 4)),
+            (Normal(-1, 2), Gumbel(-3, 4)),
+            # (Normal(-1, 2), Laplace(-3, 4)),  This case fails
```
Comment (vishwakftw) [inline test/test_distributions.py]: @fritzo I tried my best to arrive at an answer for this. Could you tell me what should be done here? If nothing works out, should I remove `_kl_normal_laplace` from the set of KL-Divergences?

File: test/test_distributions.py [hunk 2]
```diff
@@ -1209,23 +1249,99 @@ def test_laplace_shape_tensor_params(self):
...
+            (Exponential(1), Exponential(2)),
+            (Exponential(1), Gamma(2, 3)),
+            (Exponential(1), Gumbel(-2, 3)),
+            (Exponential(2), Normal(-3, 4)),
+            (Gamma(1, 2), Chi2(3)),
+            (Gamma(1, 2), Exponential(3)),
+            (Gamma(1, 2), Gamma(3, 4)),
+            (Gamma(1, 2), Gumbel(-3, 4)),
+            (Gumbel(-1, 2), Gumbel(-3, 4)),
+            (Gumbel(-1, 2), Normal(-3, 4)),  # This case fails for n = 22000
```
Comment (fritzo) [inline test/test_distributions.py]: nit: It would be a little clearer to say "This case fails for n <= 22000" here and below.

**Comment by fritzo** [comment]
@vishwakftw One of my colleagues is impressed by the breadth of implementations here and was wondering where he can find derivations. Could you point us to any literature you used?

**Comment by fritzo** [comment]
cc @karalets

**Comment by vishwakftw** [comment]
@fritzo For some of the implementations I have mentioned the links. Rest of them, I derived by hand.

File: torch/distributions/beta.py [hunk 3]
```diff
@@ -29,9 +30,10 @@ class Beta(Distribution):
     def __init__(self, alpha, beta):
         if isinstance(alpha, Number) and isinstance(beta, Number):
             alpha_beta = torch.Tensor([alpha, beta])
+            self.alpha, self.beta = torch.Tensor([alpha]), torch.Tensor([beta])
         else:
-            alpha, beta = broadcast_all(alpha, beta)
-            alpha_beta = torch.stack([alpha, beta], -1)
+            self.alpha, self.beta = broadcast_all(alpha, beta)
```
[the remark this answers is not in the document]
Comment (vishwakftw) [reply torch/distributions/beta.py]: Thank you for the comment. I will modify it appropriately. It seems redundant, so sorry about that.

[the remark this answers is not in the document]
Comment (vishwakftw) [reply torch/distributions/beta.py]: ‘.alpha’ and ‘.beta’ are required for many other KL Divergences. That is why I had to add them. I am working on trying to use ‘dirichlet’ for this.

[the remark this answers is not in the document]
Comment (fritzo) [reply torch/distributions/beta.py]: You could consider adding them as properties like
```py
@property
def alpha(self):
    return self._dirichlet.alpha.index_select(-1, 0)
```

**Comment by vishwakftw** [comment]
I have circumvented the problem in a similar manner. Once I get internet
access on my laptop, I will push these changes. Sorry about this.

File: torch/distributions/beta.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vishwakftw) [reply torch/distributions/beta.py]: I think I have circumvented this in a very similar manner. Once I get internet access in my laptop I will push the new changes.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
@vishwakftw can you please resolve the conflicts?

**Comment by vishwakftw** [comment]
@apaszke I have resolved the conflicts.

**Comment by apaszke** [comment]
Nope, still can't merge

**Comment by vishwakftw** [comment]
I think now it should be fine. Sorry. I wonder why it didn't happen earlier.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks a lot @vishwakftw!

**Pull request merged**